optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,371 @@
+                    distribution, distribution_list
+                )
+            )
+
+        assert low < high
+
+        return low, high, q
+
+    def _transform_to_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+
+            assert isinstance(distribution, _DISTRIBUTION_CLASSES)
+            if isinstance(
+                distribution,
+                (distributions.LogUniformDistribution, distributions.IntLogUniformDistribution),
+            ):
+                samples = np.log(samples)
+
+            transformed[param_name] = samples
+        return transformed
+
+    def _transform_from_uniform(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Dict[str, np.ndarray]:
+
+        transformed = {}
+        for param_name, samples in multivariate_samples.items():
+            distribution = self._search_space[param_name]
+
+            assert isinstance(distribution, _DISTRIBUTION_CLASSES)
+            if isinstance(distribution, distributions.UniformDistribution):
+                transformed[param_name] = samples
+            elif isinstance(distribution, distributions.LogUniformDistribution):
+                transformed[param_name] = np.exp(samples)
+            elif isinstance(distribution, distributions.DiscreteUniformDistribution):
+                q = self._q[param_name]
+                samples = np.round((samples - distribution.low) / q) * q + distribution.low
+                transformed[param_name] = np.clip(samples, distribution.low, distribution.high)
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                q = self._q[param_name]
+                samples = np.round(samples / q) * q
+                transformed[param_name] = np.clip(
+                    samples, distribution.low, distribution.high
+                ).astype(int)
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                samples = np.round(np.exp(samples))
+                transformed[param_name] = np.clip(
+                    samples, distribution.low, distribution.high
+                ).astype(int)
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
+
+        return transformed
+
+    def _precompute_sigmas0(self, multivariate_samples: Dict[str, np.ndarray]) -> np.ndarray:
+        # We use Scott's rule for bandwidth selection.
+        # This rule was used in the BOHB paper.
+        n_samples = next(iter(multivariate_samples.values())).size
+        n_samples = max(n_samples, 1)
+        n_params = len(multivariate_samples)
+        # TODO(kstoneriv3): The constant factor 0.2 might not be optimal.
```

Comment:
I think it is good to make this "constant" tunable. That is, how about making a new input `sigma0_magnitude` for the `parameters` in `_MultivariateParzenEstimator.__init__`.? I mean
- [1] Create `_MultivariateParzenEstimatorParameters` with parameters of the normal Parzen estimator and `sigma0_magnitude`.
- [2] Make the input of parameters in `_MultivariateParzenEstimator.__init__` as `_MultivariateParzenEstimatorParameters`.

Reply:
Good idea!

Reply:
I feel that we should include `sigma0_magnitude` in `_ParzenEstimatorParameter` as `_MultivariateParzenEstimator`'s arguments are organized followingly. What do you think?
```
class _MultivariateParzenEstimator:

    def __init__(
        self,
        multivariate_observations: Dict[str, np.ndarray],
        search_space: Dict[str, BaseDistribution],
        parameters: _ParzenEstimatorParameters,
    ) -> None:
```